### PR TITLE
Add git hooks and setup

### PR DESCRIPTION
**Pull request by ethanuppal**
Automatically `cargo clippy` before you commit

### CODE REVIEW

File: docs/intro.md [hunk 1]
```diff
@@ -49,6 +49,7 @@ Then build the compiler:
 ```
 cargo build
 ```
+Make sure to setup the git hooks via `/bin/sh setup_hooks.sh`.
```
Comment (rachitnigam) [inline docs/intro.md:52]: This should be formatted as a paragraph and we should provide some explanation about why we're asking people to do this (we enforce a formatting style in the CI so it's much easier if you set this up to make the commit process easier).

File: .hooks/pre-commit [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#/bin/sh
+
+cargo clippy
+if [ $? -neq 0 ]; then
+    exit 1
+fi
```
Comment (rachitnigam) [inline .hooks/pre-commit:6]: We also require the python files to be formatted using `flake` and `black`. Maybe worth adding them too?

**Review by rachitnigam — APPROVED** [review]
Minor comments. Resolve and merge at discretion!

File: .hooks/pre-commit (hunk 2 above)
Comment (ethanuppal) [reply .hooks/pre-commit:6]: There are tons of errors when running `$(which flake8 || which flake) .` from the root

**Pull request merged**